tests/api.c

```diff
@@ -3958,6 +3967,866 @@ static int test_wc_RipeMdFinal (void)
+    byte        hash[SHA3_512_DIGEST_SIZE];
+    byte        hashRet[SHA3_512_DIGEST_SIZE];
+
+    /* Init stack variables. */
+    XMEMSET(hash, 0, sizeof(hash));
+
+    printf(testingFmt, "wc_Sha3_512_Final()");
+
+    ret = wc_InitSha3_512(&sha3, HEAP_HINT, devId);
+    if (ret != 0) {
+        return ret;
+    }
+
+    ret= wc_Sha3_512_Update(&sha3, (byte*)msg, (word32)XSTRLEN(msg));
+    if (ret == 0) {
+        ret = wc_Sha3_512_Final(&sha3, hash);
+        if (ret == 0 && XMEMCMP(expOut, hash, SHA3_512_DIGEST_SIZE) != 0) {
+            ret = SSL_FATAL_ERROR;
+        }
+    }
+    /* Test bad args. */
+    if (ret == 0) {
+        ret = wc_Sha3_512_Final(NULL, hash);
+        if (ret == 0) {
+            ret = wc_Sha3_384_Final(&sha3, NULL);
+        }
+        if (ret == BAD_FUNC_ARG) {
+            ret = 0;
+        } else if (ret == 0) {
+            ret = SSL_FATAL_ERROR;
+        }
+    }
+    printf(resultFmt, ret == 0 ? passed : failed);
+    if (ret == 0) {
+        printf(testingFmt, "wc_Sha3_512_GetHash()");
+
+        ret = wc_InitSha3_512(&sha3, HEAP_HINT, devId);
+        if (ret != 0) {
+            return ret;
+        }
+        /* Init stack variables. */
+        XMEMSET(hash, 0, sizeof(hash));
+        XMEMSET(hashRet, 0, sizeof(hashRet));
+
+        ret= wc_Sha3_512_Update(&sha3, (byte*)msg, (word32)XSTRLEN(msg));
+        if (ret == 0) {
+            ret = wc_Sha3_512_GetHash(&sha3, hashRet);
+        }
+        if (ret == 0) {
+            ret = wc_Sha3_512_Final(&sha3, hash);
+            if (ret == 0 && XMEMCMP(hash, hashRet, SHA3_512_DIGEST_SIZE) != 0) {
+                ret = SSL_FATAL_ERROR;
+            }
+        }
+        if (ret == 0) {
+            /* Test bad args. */
+            ret = wc_Sha3_512_GetHash(NULL, hashRet);
+            if (ret == BAD_FUNC_ARG) {
+                ret = wc_Sha3_512_GetHash(&sha3, NULL);
+            }
+            if (ret == BAD_FUNC_ARG) {
+                ret = 0;
+            } else if (ret == 0) {
+                ret = SSL_FATAL_ERROR;
+            }
+        }
+
+        printf(resultFmt, ret == 0 ? passed : failed);
+    }
+
+    wc_Sha3_512_Free(&sha3);
+#endif
+    return ret;
+} /* END test_wc_Sha3_512_Final */
+
+
+/*
+ *  Testing wc_Sha3_224_Copy()
+ */
+static int test_wc_Sha3_224_Copy (void)
+{
+    int         ret = 0;
+#if defined(WOLFSSL_SHA3) && !defined(WOLFSSL_NOSHA3_224)
+    Sha3        sha3, sha3Cpy;
+    const char* msg = "Everyone gets Friday off.";
+    word32      msglen = (word32)XSTRLEN(msg);
+
+    printf(testingFmt, "wc_Sha3_224_Copy()");
+
+    ret = wc_InitSha3_224(&sha3, HEAP_HINT, devId);
+    if (ret != 0) {
+        return ret;
+    }
+
+    ret = wc_InitSha3_224(&sha3Cpy, HEAP_HINT, devId);
+    if (ret != 0) {
+    wc_Sha3_224_Free(&sha3);
+    return ret;
+    }
+
+    ret = wc_Sha3_224_Update(&sha3, (byte*)msg, msglen);
+
+    if (ret == 0) {
+        ret = wc_Sha3_224_Copy(&sha3Cpy, &sha3);
```

Comment:
Should we be verifying that the copy works correctly?  For example getting the final hash from both old and new copy and verifying they are the same value?

Same comment applies to other SHA-3 copy functions.